### DISCUSSION TITLE
Can I create prefab validation script?

### DISCUSSION THREAD

**Discussion by locus84** [TylerTemp/SaintsField#115] [Ideas]
For my company project, I created [Required] attribute.
And add a test for searching all serialized properties with that attribute are assigned in prefabs.
Maybe I can utilize saintsfield for similar purpose.(Maybe this is similar to OdinValidator)

Is there a way to create test for validating all validation attributes are met for a component or a serialized property?

**Comment by TylerTemp** [comment]
Not yet. ATM I either can't give a instruction of how it can happen ...

But the plan is: <https://github.com/TylerTemp/SaintsField/issues/98#issuecomment-2558007162>, I quote:

> Make an auto-runner for Required and auto-getters for scenes, a project, and for Addressable, with a config to allow specify assets to be checked also (I need this for my own project...)

So it'll be implemented in near feature (but no promise when tho...). Please stay tuned

**Reply by locus84** [reply]
Great to hear, but I think if the system has to meet various requirements for every project, it may be too much complicated. Just a utility function doing do job might be fine for me.

Thank you in advance.

**Reply by TylerTemp** [reply]
The logic is here: <https://github.com/TylerTemp/SaintsField/blob/4702e0c5eca488c8db4629fac46eb2f2fd04b183/Editor/Drawers/RequiredAttributeDrawer.cs#L18>

So the logic is:

1.  Read your target
2.  Convert to a serialized target: `using(var serTarget = new SerializedObect(yourTarget);`
3.  using reflection to get the `FieldInfo` which has the `Required`, also use `FindProperty` etc to get the `SerializedProperty`
4.  Call `(string curError, int _, object curValue) = SaintsField.Editor.Utils.Util.GetValue(property, field, target);`
5.  Call `bool isOk = ReflectUtils.Truly(curValue)`

**Reply by locus84** [reply]
Yeah, but there are many validation related attributes in SaintsField, like FieldType, ValidateInput, Min/MaxValue, RequireType and so on..., how's implement ITestable interface?(for example) for those attributes?

**Reply by TylerTemp** [reply]
Yeap, and auto getters. That's why I said "can't give an instruction of how it can happen"... it's a lot of tedious work and need to be made step by step. Auto Getters has a thing called "HelperArraySize" (yes, it was designed for array, I'm too lazy to change the name for that internal call) which can somewhat helping the validation. But most drawers does not have one like that. It requires some work

**Comment by TylerTemp** [comment]
Hi,

the tool for `Required` is now added in ~~[3.14.0](https://github.com/TylerTemp/SaintsField/releases/tag/3.14.0)~~ [3.14.1](https://github.com/TylerTemp/SaintsField/releases/tag/3.14.1) under  `Window` - `Saints` - `Auto Runner`

[![video](https://github.com/user-attachments/assets/bf5e7b7a-c15c-4fa4-92b9-53621d41ccb4)](https://github.com/user-attachments/assets/76683bc3-cfea-4952-9377-788e02d7e075)

More attributes support will be added later. Please stay tuned :)

**Comment by TylerTemp** [comment]
Hi,

Though not all attribute's validation are added, now in [3.18.0](https://github.com/TylerTemp/SaintsField/releases/tag/3.18.0) there is `AutoRunnerWindowBase` for easily validation which allows checking for:

1.  `Required`
2.  `ValidateInput`
3.  auto getters
4.  `OnValidate()` function in a script <--- I really need this...

`AutoRunnerWindow` is still the most featured component, but here's a simple code of how can `AutoRunnerWindowBase` be used to validate a folder (or some folders) of prefab:

```csharp
public class AutoRunnerTemplate: AutoRunnerWindowBase
{
    // Omitted code...
    
    [GetByXPath("resources::/*.prefab")]  // any prefab in the resources folders
    [GetByXPath("assets::/Config/Heros/*.asset")]  // hero scriptable objects  
    public UnityEngine.Object[] extraResources;

    // return your assets here
    protected override IEnumerable<Object> GetExtraAssets() => extraResources;

    // Omitted code...

    [Ordered, Button("Run!")]
    private IEnumerator Run()
    {
        foreach (ProcessInfo info in RunAutoRunners())
        {
            // just some info
            Debug.Log(info);

            // avoid completely frozen the editor
            if(info.ProcessCount % 100 == 0)
            {
                yield return null;
            }
        }
    }

    // at last, show the auto running results.
    // ReSharper disable once UnusedMember.Global
    [Ordered, AutoRunnerWindowResults] public List<AutoRunnerResult> ShowResults => Results;

    // important! Clean the cached results and release the loaded resources when closing the window
    public override void OnEditorDestroy()
    {
        CleanUp();
    }
}
```

**Reply by locus84** [reply]
This is great!!, now I can use this for my ci/cd.
going to check the functionality asap!!
